nifi/nifi-nar-bundles/nifi-kite-bundle/nifi-kite-processors/src/main/java/org/apache/nifi/processors/kite/ConvertAvroSchema.java

```diff
@@ -0,0 +1,252 @@
+					"Cannot find field mapping: " + fieldMappingProperty);
+			session.transfer(incomingAvro, FAILURE);
+			return;
+		}
+		final AvroRecordConverter converter = new AvroRecordConverter(
+				inputSchema, outputSchema, fieldMapping);
+
+		final DataFileWriter<Record> writer = new DataFileWriter<>(
+				AvroUtil.newDatumWriter(outputSchema, Record.class));
+		writer.setCodec(CodecFactory.snappyCodec());
+
+		try {
+			final LongHolder written = new LongHolder(0L);
+			final FailureTracker failures = new FailureTracker();
+
+			FlowFile badRecords = session.clone(incomingAvro);
+			FlowFile outgoingAvro = session.write(incomingAvro,
+					new StreamCallback() {
+						@Override
+						public void process(InputStream in, OutputStream out)
+								throws IOException {
+							try (DataFileStream<Record> stream = new DataFileStream<Record>(
+									in, new GenericDatumReader<Record>(
+											converter.getInputSchema()))) {
+								try (DataFileWriter<Record> w = writer.create(
+										outputSchema, out)) {
+									for (Record record : stream) {
+										try {
+											Record converted = converter
+													.convert(record);
+											w.append(converted);
+											written.incrementAndGet();
+										} catch (AvroConversionException e) {
+											failures.add(e);
+											getLogger().error(
+													"Error converting data: "
+															+ e.getMessage());
+										}
+									}
+								}
+							}
+						}
+					});
+
+			long errors = failures.count();
+
+			// update only if file transfer is successful
+			session.adjustCounter("Converted records", written.get(), false);
+			// update only if file transfer is successful
+			session.adjustCounter("Conversion errors", errors, false);
+
+			if (written.get() > 0L) {
+				session.transfer(outgoingAvro, SUCCESS);
+
+				if (errors > 0L) {
+					getLogger().warn(
+							"Failed to convert {}/{} records from CSV to Avro",
+							new Object[] { errors, errors + written.get() });
+					badRecords = session.putAttribute(badRecords, "errors",
+							failures.summary());
+					session.transfer(badRecords, INCOMPATIBLE);
+				} else {
+					session.remove(badRecords);
+				}
+
+			} else {
+				session.remove(outgoingAvro);
+
+				if (errors > 0L) {
+					getLogger().warn(
+							"Failed to convert {}/{} records from CSV to Avro",
+							new Object[] { errors, errors });
+					badRecords = session.putAttribute(badRecords, "errors",
+							failures.summary());
+				} else {
+					badRecords = session.putAttribute(badRecords, "errors",
+							"No incoming records");
+				}
+
+				session.transfer(badRecords, FAILURE);
+			}
+
+		} catch (ProcessException | DatasetIOException e) {
+			getLogger().error("Failed reading or writing", e);
+			session.transfer(incomingAvro, FAILURE);
+		} catch (DatasetException e) {
+			getLogger().error("Failed to read FlowFile", e);
+			session.transfer(incomingAvro, FAILURE);
+		} finally {
+			try {
+				writer.close();
```

Comment:
I think this is handled by the try-with-resources block in the `StreamCallback`. The writer should be closed correctly after that so there is no need to do it here.